### PR TITLE
better convergence of compute_jumps

### PR DESCRIPTION
**Pull request by SnoopyLane**
TLDR: Two passes of compute_jumps() is not sufficient.

Consider the following sequence of instructions (assuming Python 3.6 / wordcode):

> Label1:
>     JUMP_ABSOLUTE Label2
>     ...126 instructions...
> Label2:
>     NOP

On first pass of `compute_jumps()`, Label2 will be at address 254, so that value encodes into the single byte arg of JUMP_ABSOLUTE.  All is well.

Now imagine instead that one of the intervening "...126 instructions..." needed to have an EXTENDED_ARG prepended to it.  So the initial pass of compute_jumps() will return True (i.e. modified).  So we call compute_jumps() again...

Now during this second pass of compute_jumps(), Label2 is seen to be at address 256, so the offset from "JUMP_FORWARD Label2" is 256 and cannot be encoded in a single byte, so...  we prepend EXTENDED_ARG, remember that we modified that code, return True, and then...

The calling code in `to_concrete_bytecode()` raises if the second compute_jumps() also has modifications.

Clearly more than two passes can be necessary.

You could arrange this scenario pre-wordcode as well, but since the jumps have 16 bit offsets you need a much larger function before this becomes a problem.

Given the current algorithm of just calling compute_args() until it converges, it needs to be called more than twice.  I'm not sure what the upper limit could be.  But I'm not comfortable with it iterating forever, so I changed the upper limit to 10 passes before it raises.

I think ultimately this should probably have a better algorithm that doesn't need to iterate over every instruction until it converges on a solution, but I haven't read compile.c yet to see how it's done there.

### CODE REVIEW

**Comment by MatthieuDartiailh** [comment]
I think I agree with your analysis here, but I would like to see a test for this (even if it is limited to wordcode). Could you add one ?

**Comment by SnoopyLane** [comment]
Not sure how you feel about my addition of _ConvertBytecodeToConcrete.max_compute_jumps_steps.  I prefer to test the failure case, and without polluting the public API with hooks to set up the failures.

File: bytecode/concrete.py [hunk 1]
```diff
@@ -498,6 +498,8 @@ def compute_jumps(self):
 
         return modified
 
+    max_compute_jumps_steps = 10
```
Comment (MatthieuDartiailh) [inline bytecode/concrete.py]: Could you please:
- make this variable private ?
- define it right after the class definition before the first method ?
- add a comment describing its purpose with a link to the relevant part of the CPython source ?

File: bytecode/concrete.py [hunk 2]
```diff
@@ -506,12 +508,12 @@ def to_concrete_bytecode(self):
...
-        if modified:
+        for step in range(0, self.max_compute_jumps_steps):
             modified = self.compute_jumps()
-            if modified:
-                raise RuntimeError("compute_jumps() must not modify jumps "
-                                   "at the second iteration")
+            if not modified:
+                break
+        else:
+            raise RuntimeError("compute_jumps() failed to converge")
```
Comment (MatthieuDartiailh) [inline bytecode/concrete.py]: It may be interesting to indicate after how many iteration we stopped.

**Comment by serhiy-storchaka** [comment]
Wouldn't be problems with a code like
```python
script = "if x: x\n%selse: %r" % ("elif x: x\n"*100, list(range(128)))
```
?

**Comment by MatthieuDartiailh** [comment]
@serhiy-storchaka I am not sure it is related. The point of this code is to compute the offset for the relative jumps, that requires multiple passes simply because in some cases the addition of EXTENDED_ARG because some jumps are large can modify the offsets. As a consequence (and as noted in CPython source (https://github.com/python/cpython/blob/master/Python/compile.c#L5274)), it is expected to converge quickly.

@SnoopyLane could you test @serhiy-storchaka case, out of curiosity ?

**Comment by SnoopyLane** [comment]
I agree with your review comments.  But before I push yet another change, let me ask this...

Since compile.c has no limit (as you've pointed out twice now :-), what if I remove the limit from my changes?  Frankly, I'm not fond of that, I think there should be a limit in compile.c.  But git blame says that has been there for at least years.

Removing any limit has the upside of matching the logic in compile.c (arguably just as bad as good), but also simplifying my change to concrete.py, simplifies the test, and of course removes the need to expose a way for the tests to hammer the behavior.

Also, I'm not sure what the theoretical worst case could be.  I'm not convinced that it's not something on the order of (2**offset) / sizeof(jump_instruction).  I.e. 256/2 on wordcode, or 65536/3 on bytecode.  But I confess that I haven't spent time thinking about that case, nor trying to find old discussion around the code in compile.c.   I think, just for fun, I shall attempt to build a bad case later today.

**Comment by MatthieuDartiailh** [comment]
I am not a big fan of the idea of having absolutely no limit. Furthermore sometimes people do surprising things when modifying the bytecode and can encounter cases rarely seen in usual Python code, which could infinitely loop. So I believe it is reasonable to have a limit (in addition we do expect that the process converges quickly in most cases). 
However if we worry that we may have set it too low, we could add a keyword-only argument to `to_concrete_bytecode` to let people easily change the limit. In which case the detailed explanation could be a note in the docstring.

**Comment by SnoopyLane** [comment]
I made changes addressing your review comments, but no more.

Before raising the number, or allowing it to be changed as you suggest, I want to spend some time investigating not just worst theoretical case but also some practical examples.

File: bytecode/concrete.py (hunk 1 above)
Comment (SnoopyLane) [reply bytecode/concrete.py]: done

File: bytecode/concrete.py (hunk 2 above)
Comment (SnoopyLane) [reply bytecode/concrete.py]: good idea; done

**Comment by MatthieuDartiailh** [comment]
Make sense. Keep me posted about your progress.

**Comment by SnoopyLane** [comment]
FYI:  I have a new test the demonstrates that there is no limit to the number of passes.  It's limited only by the size of the input code.  I'm going to follow your suggestion of making the number of loops controllable via keyword argument, then push those changes.

I'll also change the default number to something smaller than I currently have (129).  

For an additional datapoint, I hacked my copy of cpython to record the number of iterations needed in compile.c, then rm'd all the .pyc files, then run cpython's unittests.  A few tests needed 5 passes, but that was the maximum.  Not that unittests are a torture-test for this kind of thing, but I thought it an interesting datapoint.

**Comment by SnoopyLane** [comment]
I just pushed a set of changes that provide:
* Test that requires 385 iterations of compute_jumps.  I think this serves as demonstration that the number of passes is bounded only by the input size.
* Control the limit via keyword argument as you suggested earlier.
* Doc changes for the keyword argument (separate commit).

I only just saw your comment about documenting the 5 passes needed for unit tests.  Let me ponder the wording for a bit.  That number "5" certainly isn't contractual and I don't want the docs to suggest to somebody that they should base any decisions on that.

**Comment by MatthieuDartiailh** [comment]
I gave a quick look at your changes and they look really good. The number five may be added in the api docs when you mention that it converges quickly most of the time and give the example of the Python test suite.

**Comment by SnoopyLane** [comment]
Final doc tweak pushed with 172199ab.  Unless your review spots problems I would like to put this to get this merged an move on.

File: doc/api.rst [hunk 3]
```diff
@@ -348,20 +348,32 @@ Bytecode
...
+   .. method:: to_concrete_bytecode(compute_jumps_passes: int = None) -> ConcreteBytecode
 
       Convert to concrete bytecode with concrete instructions.
 
       Resolve jump targets: replace abstract labels (:class:`Label`) with
-      concrete instruction offsets (relative or absolute, depending on the jump
-      operation).
+      concrete instruction offsets (relative or absolute, depending on the
+      jump operation).  It will also add EXTENDED_ARG prefixes to jump
+      instrctions to ensure that the target instructions can be reached.
```
Comment (MatthieuDartiailh) [inline doc/api.rst]: instrctions -> instructions  other than that this is good to merge !

**Comment by SnoopyLane** [comment]
Not sure how I missed that in the spell checker.

**Comment by MatthieuDartiailh** [comment]
No problem I will merge later today. Thanks again for your great work on this.
